flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,226 @@
+  ///
+  /// Stopping the frame tracking means setting [isTrackingPaused] is `true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isTrackingPaused = false;
+
+    if (!_isTrackingRegistered) {
+      _frameCallbackHandler.addPersistentFrameCallback(measureFrameDuration);
+      _isTrackingRegistered = true;
+    }
+  }
+
+  /// Records the duration of a single frame and stores it in [frameDurations].
+  ///
+  /// This method is called for each frame when frame tracking is active.
+  Future<void> measureFrameDuration(Duration duration) async {
+    // Using the stopwatch to measure the frame duration is flaky in ci
+    if (_isTestMode) {
+      // ignore: invalid_use_of_internal_member
+      frameDurations[options.clock()] = duration.inMilliseconds;
+      return;
+    }
+
+    if (_isTrackingPaused) return;
+
+    if (!_stopwatch.isRunning) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler.endOfFrame;
+
+    final frameDuration = _stopwatch.elapsedMilliseconds;
+    // ignore: invalid_use_of_internal_member
+    frameDurations[options.clock()] = frameDuration;
+
+    _stopwatch.reset();
+
+    if (_frameCallbackHandler.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  void _applyFrameMetricsToSpan(
+      ISentrySpan span, Map<String, int> frameMetrics) {
+    frameMetrics.forEach((key, value) {
+      span.setData(key, value);
+    });
+
+    // This will call the methods on the tracer, not on the span directly
+    if (span is SentrySpan && span.isRootSpan) {
+      frameMetrics.forEach((key, value) {
+        // ignore: invalid_use_of_internal_member
+        span.tracer.setData(key, value);
+      });
+      frameMetrics.forEach((key, value) {
+        // In measurements we change e.g frames.total to frames_total
+        // We don't do span.tracer.setMeasurement because setMeasurement in SentrySpan
+        // uses the tracer internally
+        span.setMeasurement(key.replaceAll('.', '_'), value);
+      });
+    }
+  }
+
+  @visibleForTesting
+  Map<String, int> calculateFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp, int displayRefreshRate) {
+    final expectedFrameDuration = ((1 / displayRefreshRate) * 1000).toInt();
+
+    // Filter frame durations within the span's time range
+    final timestamps = frameDurations.keys
+        .takeWhile((value) =>
+            value.isBefore(endTimestamp) && value.isAfter(span.startTimestamp))
+        .toList();
+
+    if (timestamps.isEmpty) {
+      options.logger(
+          SentryLevel.info, 'No frame durations available in frame tracker.');
+      return {};
+    }
+
+    int slowFramesCount = 0;
+    int frozenFramesCount = 0;
+    int slowFramesDuration = 0;
+    int frozenFramesDuration = 0;
+    int framesDelay = 0;
+
+    for (final timestamp in timestamps) {
+      final frameDuration = frameDurations[timestamp] ?? 0;
```

Comment:
The frames should be removed from the map, so they are not counted in other spans
```suggestion
      final frameDuration = frameDurations.remove(timestamp) ?? 0;
```

Reply:
In a scenario like this: root span (t: 1-10) and its child span (t: 6-8).

If we remove frame durations after calculating metrics for the child span, 
wouldn't this leave gaps in the data needed for the root span's calculations?

Should the root span have access to all frame durations within its time range, 
regardless of whether they've been used in child span calculations?

or maybe I misunderstood how we need to calculate it

Reply:
btw frames are also being removed if active/running spans is not empty
```dart
    activeSpans.remove(span);
    if (activeSpans.isEmpty) {
      clear();
    } else {
      frames.removeWhere(
          (frameTimestamp, _) => frameTimestamp.isBefore(activeSpans.first.startTimestamp));
    }
```

Reply:
> Shouldn't the root span have access to all frame durations within its time range,
regardless of whether they've been used in child span calculations?

I recall we said to set them only on the current span, but the i found we have to set to [all concurrent spans](https://develop.sentry.dev/sdk/performance/frames-delay/), so it's good as is